#!/bin/bash

# The cellranger vdj pipeline can be used to analyze sequencing data produced from Chromium Next GEM Single Cell 5 prime V(D)J libraries. 
# It takes FASTQ files from cellranger mkfastq or bcl2fastq for V(D)J libraries and performs sequence assembly and paired clonotype calling. 
 
# Important Arguments
# --id: Required. A unique run ID string (e.g., Sample1_Fastq).
# --fastqs: Required. Path of the fastq_path folder generated by cellranger mkfastq.
# --transcriptome: Required. Path to the Cell Ranger-compatible transcriptome reference e.g.,
# For a human-only sample, use /opt/refdata-cellranger-GRCh38-3.0.0

# https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/using/vdj

# Note: this requires high resources (ram and processing core).
cellranger vdj --id=Sample1_VDJ_outPut \
                 --reference=/opt/refdata-cellranger-vdj-GRCh38-alts-ensembl-5.0.0 \
                 --fastqs=/Functional_Heterogeneity/VDJ/fastq_path \
                 --sample=Sample1_Fastq \

