#!/bin/bash

# Bash Script to generate single cell feature counts for a single library using cellranger count.
# Cellranger count takes raw fastq files as input and returns feature-barcode matrices.

# Important Argument
# --id: Required. A unique run ID string (e.g., Sample1_Fastq).
# --fastqs: Required. Path of the fastq_path folder generated by cellranger mkfastq.
# --transcriptome: Required. Path to the Cell Ranger-compatible transcriptome reference e.g.,
# For a human-only sample, use /opt/refdata-cellranger-GRCh38-3.0.0

# https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/count#cr-count

# Note: this requires high resources (ram and processing core).
# Run cellranger count to generate single cell feature counts.
# Required. Sample name as specified in the sample sheet supplied to cellranger mkfastq.

cellranger count --id=Sample1_GEX_outPut \
                   --transcriptome=/opt/refdata-cellranger-GRCh38-3.0.0 \
                   --fastqs=/Functional_Heterogeneity/GEX/Sample1_Fastq \
                   --sample=Sample1_Fastq \
